collector/exchange.go

```diff
@@ -0,0 +1,679 @@
+		OWARequestsPerSec:  desc("owa_requests_per_sec", "Number of requests handled by Outlook Web App per second"),
+
+		// MS Exchange Autodiscover
+		AutodiscoverRequestsPerSec: desc("autodiscover_requests_per_sec", "Number of autodiscover service requests processed each second"),
+
+		// MS Exchange Workload Management
+		ActiveTasks:    desc("workload_active_tasks", "Number of active tasks currently running in the background for workload management"),
+		CompletedTasks: desc("workload_completed_tasks", "Number of workload management tasks that have been completed"),
+		QueuedTasks:    desc("workload_queued_tasks", "Number of workload management tasks that are currently queued up waiting to be processed"),
+	}
+
+	collectorFuncLookup := map[string]collectorFunc{
+		"ldap":                 c.collectLDAP,
+		"transport_queues":     c.collectTransportQueues,
+		"database_instances":   c.collectDatabaseInstances,
+		"http_proxy":           c.collectHTTPProxy,
+		"activesync":           c.collectActiveSync,
+		"availability_service": c.collectAvailabilityService,
+		"owa":                  c.collectOWA,
+		"autodiscover":         c.collectAutoDiscover,
+		"management_workloads": c.collectManagementWorkloads,
+		"rpc":                  c.collectRPC,
+	}
+
+	// get the disabled and enabled collectors into slices
+	disabledCollectors := strings.Split(*argExchangeDisabledCollectors, ",")
+	enabledCollectors := strings.Split(*argExchangeEnabledCollectors, ",")
+
+	// collFuncNames that are not also disabledCollectorFuncs gets added to the ActiveCollFuncs slice.
+	for _, collFuncName := range enabledCollectors {
+		collFunc, isValidName := collectorFuncLookup[collFuncName]
+
+		if !isValidName {
+			return nil, fmt.Errorf("No such collector function %s", collFuncName)
+		}
+
+		// skip collector func names that are explicitly disabled
+		if contains(disabledCollectors, collFuncName) {
+			continue
+		}
+
+		c.ActiveCollFuncs = append(c.ActiveCollFuncs, collFunc)
+	}
+
+	if *argExchangeListAllCollectors {
+		state := ""
+		for _, name := range exchangeAllCollectorFuncs {
+			if contains(disabledCollectors, name) {
+				state = "[disabled] "
+			}
+
+			if contains(enabledCollectors, name) {
+				state = "[enabled] "
+			}
+
+			fmt.Printf("%-15s %-32s %-32s\n", state, name, exchangeCollectorFuncDesc[name])
+		}
+	}
+
+	return &c, nil
+}
+
+// Collect collects exchange metrics and sends them to prometheus
+func (c *exchangeCollector) Collect(ctx *ScrapeContext, ch chan<- prometheus.Metric) error {
+	for _, collFunc := range c.ActiveCollFuncs {
+		if err := collFunc(ch); err != nil {
+			log.Errorf("Error in %s: %s", className(collFunc), err)
+		}
+	}
+	return nil
+}
+
+func (c *exchangeCollector) collectLDAP(ch chan<- prometheus.Metric) error {
+	data := []win32_PerfRawData_MSExchangeADAccess_MSExchangeADAccessProcesses{}
+	if err := wmi.Query("SELECT * FROM "+className(data), &data); err != nil {
```

Comment:
This pattern is a bit different from other WMI collectors, where we use `queryAll(&data)` (which basically does what you do here). Is there a reason not to?

Reply:
Maybe it served some purpose before, but it doesn't look like it anymore. 😄I'll update it